### DISCUSSION TITLE
vi_server_for_subvi

### DISCUSSION THREAD

**Discussion by giscus** [ruanqizhen/labview_book#52] [Announcements]
# vi_server_for_subvi

静态与动态装载子 VI

https://lv.qizhen.xyz/vi_server_for_subvi

**Comment by GRIT-LRJ** [comment]
老师您好，在学习异步调用的例子时调用“开始异步调用”时，引用处连线报错：“通过引用调用”函数的所有类型的引用输入必须为严格VI引用，请问老师这个应该怎么解决

**Reply by ruanqizhen** [reply]
看你的错误提示是要求把引用设置为严格类型，上文中倒数第二段说的就是这个问题，你试试看照着做行不行？

**Reply by GRIT-LRJ** [reply]
可以了，哇，谢谢老师！

**Reply by GRIT-LRJ** [reply]
老师 我又有问题了，使用“开始异步调用”函数后的程序点击停止按钮后，仪表还是在转动，这是什么原因呢。
另外还有一个小小的问题就是仪表如何设置0刻线和10重合。

**Reply by ruanqizhen** [reply]
设置仪表的刻度线比较容易，把鼠标挪到刻度线上，在合适的位置就可以点击拖动刻度线进行调整了。
仪表不停下，应该是因为被调用的子程序没有停止。在正文里搜索“检测主程序中 Stop 按钮”，参考这一段的代码。

**Reply by GRIT-LRJ** [reply]
老师我找到原因了，谢谢老师！

**Comment by GRIT-LRJ** [comment]
老师您好，想向老师请教一个问题：怎么在主vi中实现修改异步调用vi的输入控件值呢。

**Reply by ruanqizhen** [reply]
如果 VI 要修改同 VI 中的控件的值，就使用数据线传入数据就可以。或者有时候也可以通过局部变量修改。
如果 VI 要修改另一个 VI 中的控件的值，那么需要通过控件的引用，调用控件的“值”方法进行修改。

**Comment by denneswang** [comment]
老师您好：关于异步调用vi的释放， 看到一个文章，建议““打开 VI 引用”这个函数的“VI指定类型”参数，不使用 "严格类型的 VI 引用" 。是下面这个网址：https://labviewcoder.com/2016/07/06/quick-tip-asynchronously-launching-vis-the-right-way/。
这样可以能够正确的释放异步调用的vi，不知道是否正确。我在使用异步调用的时候， 能遇到这种不能正确释放的问题，但是他说的方法我没有做出来。就是“VI指定类型”参数用生成常数或者控件，都不对。您如果方便，请您指点。谢谢

**Reply by ruanqizhen** [reply]
我读了这个帖子，他说的 bug 可能是存在的，但是我自己没有遇到过。帖子是六年多前发的，不知道是不是 LabVIEW 已经修改了 bug。
有一点文章中说的是对的，“打开 VI 引用”函数只需要接线板的信息，不使用“严格类型的 VI 引用”也可以。使用“严格类型的 VI 引用”的好处是：它与 VI 是自动关联的，如果被调用的 VI 接线板被修改了，“严格类型的 VI 引用”和异步调用函数也会自动更新。使用接线板类型的 VI 引用还要手动更新。
我的建议是，可以观察一下有没有遇到他提到的 bug。如果遇到了，就采用他的方案修改。如果没有 bug，就采用 LabVIEW 的默认方式就好了。

**Reply by denneswang** [reply]
老师您好，感谢您的回复，非常详细的解释了这个问题，我再多试试。有问题再来求教

**Comment by tetsuoyyx** [comment]
阮老师，使用动态调用的时候是不是就蕴含了【每次调用是重新加载】，还是说仍然可以设置【每次调用是重新加载】or【首次调用后保留】呢？

**Reply by ruanqizhen** [reply]
动态调用时会打开一个  VI 引用，如果  VI 引用被关闭，VI 就会被卸载出内存。所以，你可以每次运行后都关闭 VI 引用，那就相当于“每次调用是重新加载”；也可以运行 VI 后，不关闭引用，下次直接使用同样的引用运行 VI，那就相当于“首次调用后保留”。

**Comment by tetsuoyyx** [comment]
阮老师，请教一下，异步调用的VI如何通过编程在调用它的VI里面把它终止运行（前面板也同时·关掉）呢？

**Reply by ruanqizhen** [reply]
使用 Abort VI 方法可以终止运行一个 VI， FB.Close 方法可以关闭前面板。你可以参考一下下面这两个部分的内容：
<https://lv.qizhen.xyz/vi_server_for_subvi#后台任务>
<https://lv.qizhen.xyz/vi_server_for_subvi#启动画面>

**Comment by tetyangy** [comment]
阮老师请教一下，动态调用的话，必须要提供路径和VI的源文件吗？不能嵌入到应用程序的EXE里面吗？
静态VI引用是不是就可以不提供被打开的vi的源文件（已经嵌入到了exe里面）？

**Reply by ruanqizhen** [reply]
这个问题我有点不太确定了，很长时间没用过了，需要找时间验证一下。不过，直觉上我觉得，动态调用也应该可以使用嵌入到 exe 中的 VI ，只是 VI 放在 EXE  中，路径会有一些变化，调用时候需要注意。可以参考一下 <https://lv.qizhen.xyz/manage_release#文件路径的变化>

**Comment by ruanqizhen** [comment]
之前博客上的相关文章，搬到这里：

### 后台任务


有一些任务在运行的时候，并不需要与用户交互。它们通常在不打扰用户其它工作的时候默默的执行。这样的任务叫后台任务。与之对比，前台任务就是用户看得见的。 比如，一个文本编辑软件，帮助用户上编辑文字的任务都是前台执行的。如果这个软件做得好，最好还有自动保存文件的功能，这样在系统突然崩溃之后，不至于所有用户做的改动都白费。自动保存这个功能一般都被作为后台任务，他在执行时，不打扰用户其它工作。 后台任务的应用还有很多。在 LabVIEW 中实现后台任务的方法，就是使用 VI Scripting 技术，动态把负责后台的任务在新的线程里启动起来。下面以一个比较常用的功能来说明后台任务是如何实现的。

大型 LabVIEW 程序启动的时候可能会比较慢，这是因为程序在它所有的子VI都被装入内存之后才能执行。把大量子VI装入内存可能要花费几秒钟的时间，如果这段时间，什么都不显示给用户，用户可能会以为是程序出错或者死机了，从而做出一些错误的操作。好的做法是，在启动程序的时候立刻显示给用户一个提示界面，告诉用户程序正在装载，已经装载的进度。同时把所有的子VI装入内存，然后再启动程序的主界面。 在程序显示启动画面这个阶段，当子VI全部装入内存后，启动画面仍然显示在前台，而程序会再启动一个后台任务，把主程序运行起来。之后，启动画面程序把自己关闭，而主程序界面也从后台跳到前台来了。 启动画面的程序代码如下图所示：



它由5个部分组成： 第1部分负责把主程序的子VI都装入内存（用 Open VI Reference 函数打开这些子 VI 就可以把它们装入内存）。这部分与后台任务无关，不细说了。 第2部分 Open VI Reference 函数打开主程序的主VI，它的输入是主VI的地址。后续使用到的针对这个 VI 的 Scripting 函数，都依赖于在此打开的 VI 的引用。 第3部分设置主VI的 Front Panel:Open 属性。这个属性值为真的时候，主程序的界面被显示出来。在我们这个程序中，主程序最终是要跳到前台来与用户交互的，所以需要显示他的界面。有些后台任务，比如自动保存任务，没有交互界面，就不需要设置此属性。 第4部分调用主VI的 Run VI 方法。调用这个方法之后，主程序开始运行。目前主程序还是作为后台任务在运行。这里需要注意的是 Run VI 方法的两个参数： Wait Until Done 参数如果设为真，启动画面程序运行到这里就会暂停下来，一直等到后台任务结束，才继续执行后面的代码。但是，作为启动画面程序，我们显然不希望出现这种情况。我们需要让前台任务和后台任务同时执行，所以这个参数在本程序中被设为“假”。后台程序在另一个线程中被执行，它的结束与否不影响当前程序的执行，我们的启动画面程序会立即去执行下一条语句。 用 Open VI Reference 函数打开的引用必须被关掉，否则，VI被装入内存，却从来不被卸除出内存，就会引起内存泄漏。Auto Dispose Ref 参数用于决定由谁来关闭动态运行起来的VI的引用，是前台任务，还是后台任务自己。真对于我们的例子，后台任务是主程序，我们希望用户关闭它的同时，把它从内存中卸除。所以实例中这个参数的值是“假”，后台程序结束时自动释放引用。对于那些没有界面的后台任务，它们的启动、停止都需要由前台程序来控制。这是这个参数就需要被设为“真”。前台程序还需要这个引用来停止后台任务、卸除后台任务的VI。 第5部分是对启动画面VI自身进行操作的。程序运行到这里，主程序已经运行起来，而启动画面程序的使命也已完成，应当让位给主程序了。启动画面程序在此调用 Front Panel:Close 把自己关闭。

**Reply by ruanqizhen** [reply]
之前博客下的留言：

Jun
2009年11月20日 08:29 编辑
谢谢！我正在尝试使用Windows API来编写服务程序

回复
Qizhen
2009年11月19日 08:51 编辑
LabVIEW本身的一些服务程序并不是使用LabVIEW G代码编写的，他们可能是使用C++编写的。

回复
Jun
2009年11月18日 13:15 编辑
那么像LabVIEW本身的一些服务程序，如NI Service Locator这种是怎么实现的呢？

回复
Qizhen
2009年11月18日 12:52 编辑
不出现在进程列表恐怕不行。我没做过Windows后台服务，但他肯定需要比一般程序高的权限，LabVIEW不一定能做到。数据交换使用共享变量最方便。

回复
Jun
2009年11月18日 09:15 编辑
谢谢！我尝试一下您的方法，不知道数据交互上是否方便。原本我很像实现那种像Windows后台服务那样的程序，不想让它出现在进程列表中。一直没有找到这方面的相关介绍和资料。

回复
Qizhen
2009年11月18日 09:11 编辑
有一个简单的办法，你编写一个没有界面的程序，制作成EXE文件。然后在注册表中把这个程序设置为开机即运行就可以了。

回复
Jun
2009年11月18日 09:08 编辑
那请问一下可以用LabVIEW来编写Windows中的常驻后台服务程序吗？

回复
Qizhen
2009年11月17日 11:12 编辑
这两种后台任务的层次不同。Windows中的常驻后台服务程序，是一个独立的程序，可以脱离任何其他程序独自运行。LabVIEW中的后台任务是LabVIEW进程的一部分。它不能脱离LabVIEW或者是LabVIEW生成的应用程序，它只能在这些应用程序中处理界面之外的工作。

回复
Jun
2009年11月17日 09:55 编辑
请问一下可以使用这个方法实现类似于Windows中的常驻后台服务程序吗？

回复
Qizhen
2009年03月11日 17:17 编辑
应该这么说，我说的比较简单，只是从使用方法上说了一下两者个区别。LV help解释的比较深入。不过后台程序界面是否打开，与这个参数无关。这个参数主要还是用来确定谁来管理被打开的VI refunm。如果设成false，必须使用Close函数来关闭被打开的VI refnum。如果设成true，那可以不调用VI refnum。当后台程序停止运行，则VI refunm自动关闭。如果后台程序有界面，则一般是关闭界面时，程序结束，VI refunm自动关闭。

回复
gufei
2009年03月8日 11:54 编辑
依您的意思，如果我动态打开的是后台程序（不需开界面）就把Auto Dispose Ref 设为true, 如果动态打开的是有界面的程序，就把Auto Dispose Ref 设为false? 那动态打开的程序所要的数据内存空间是按什么方式回收的呢？

回复
gufei
2009年03月8日 11:48 编辑
Auto Dispose Ref ： Allows you to run a VI independently of the caller without opening its front panel immediately or opening another reference inside the target VI. If TRUE, the referenced VI transfers ownership of the reference from the calling VI to the VI that is running. This means LabVIEW disposes of the reference, along with the parallel data space, when the target VI goes idle, not when the VI that opened the reference goes idle. The reference can still be used by the calling VI until the target VI closes the reference. The calling VI does not need to close the reference unless the Run VI method returns an error. If the calling VI does close the reference, the target VI can abort and leave memory. If FALSE, LabVIEW automatically disposes of the VI reference when the VI that opened it goes idle. The default is FALSE. Note If Auto Dispose Ref is TRUE and the method returns an error, LabVIEW does not transfer ownership of the reference to the target VI. LabVIEW will not automatically dispose of the reference when the target VI goes idle.

回复
gufei
2009年03月8日 11:48 编辑
可否详细解释下Auto Dispose Ref, 我感觉您说的跟LabVIEW Help 怎么不太一样啊？

回复
aragorn
2008年04月23日 11:03 编辑
這個部落格對於labview的說明
真是使我獲益良多
謝謝你

回复
果果
2008年03月5日 23:23 编辑
谢谢。

回复
agilentvee
2008年03月5日 11:49 编辑
小弟补充一点:第4步中wait until done参数为真,一般是在测试系统里面用.把很多测试subvi放在一个文件夹main函数从中调用,这样好处是可以拓展测试程序而不必更改main函数,不过要求所有subvi必须作成一种格式

回复
果果
2008年02月29日 20:38 编辑
我明白了，谢谢阮大侠了，真是好人！
高杰先生，也谢谢你的回答，我认识你一个朋友：王磊。

回复
Qizhen
2008年02月29日 17:31 编辑
用图中标号2的 Open VI Reference 函数打开一个VI的引用，就可以把这个VI连同他所有的子VI全部装入内存。如果启动画面不需要什么动态的信息，直接打开主VI的引用就行了。如果还需要显示装载的进度，你肯定能就不希望一次装一个主VI进来，而是把几百个子 VI 一个一个装入内存，没装入一个启动画面上的进度条前进一格。可以这样实现： 先得到所有子VI文件的路径。这个过程的策略可以你自己制定，比如你规定所有子VI都放在某一文件夹下，那么遍历这个文件夹，得到他下面所有 VI 文件的路径，使用 Open VI Reference 函数把它们一一打开，没打开一个，更新一下进度条。

回复
杰
2008年02月29日 15:02 编辑
静态调用都是自动加载的。动态调用，嘛玩意要调那么多啊？at least，初始化的VI里面，哪需要那么多啊！动态调用VI，如何调用和被调用多少，好像没什么太大的关系。当然，可能是我没有理解这个朋友的问题，也可能是这个朋友没表述的很清楚。譬如说，你可能想问，有什么方式调用很多子VI，比较省内存之类的。呵呵
阮虾有点忙，俺越俎代庖一下。说的不对的地方，阮虾来纠偏一哈子。

回复
果果
2008年02月28日 23:15 编辑
想问你一个小问题：装载子VI时，如果有数百个子VI，该如何装载？

回复
杰
2008年02月28日 19:20 编辑
啊！
我的程序也是这样，思路完全一致。虽然我做的程序不是很大。但是很多对仪器仪表的初始化，以及仪器仪表本身的Self-test比较费时。所以程序刚开始运行的时候，一定要做这么一个东西。而且做了这个东西，把自己公司、单位甚至自己的个人的信息放在里面，嘿嘿，感觉还是很正式。后来俺又做了改进，显示公司以及部门的log之后，详细的告诉用户，现在是在对系统的哪一个部分进行初始化。好像在放幻灯片一样，呵呵。不过，说老实话，太花哨了，我反而不喜欢。

回复
arrowpig
2008年02月26日 09:15 编辑
看到这几个函数调用好亲切啊。后来我们几乎所有的Wizard程序几乎都是用的这个思路。

回复

**Reply by ruanqizhen** [reply]
之前博客下的留言：

彪
2010年03月21日 16:16 编辑
请大家帮我分析一个LabVIEW范例程序：write table to XL.vi

回复
彪
2010年03月21日 16:16 编辑
写作背景： 1、 吃过高中学习生涯的家常便饭——连续半个月的卖命补课，在补课过程中有 12 位新同学的加入，改变了我以往对同学学习态度的看法。 2、 经过一学期的努力，以前只会说 ladies and gentlemen 的我，这次期末考试成绩名列前茅。在这个过程中和英语老师结下深厚的友谊。 3、 连续写了三次关于电源的解决方案，却被老师们砍得死去活来，好不快活啊！体会到做初期的技术开发是非常困难的。 4、 和一位 20 出头的钻井小弟畅谈他的职业生涯，惊奇发现在不同行业中也有不平凡的拼搏与坚持。 虽然我不是大名鼎鼎的文学大师，也不是语文成绩顶尖的学生，但有感而发动一动笔杆子，说不定会奇迹出现，此文献给所有正在为自己的目标而奋斗的人，相信它会激励每一个人朝着明确的方向前进。 用真诚和务实发挥影响力，经过半年的寒窗苦读之后，发现周围的环境开始和自己变得融洽起来，虽然我不敢肯定所有人都愿意培养自己的专研精神，但至少在懂事的人群中，相信他们愿意和我分享在技术上每一步的成功和失败。因为我的科任老师被我的举动所感动，决定尽力地帮助我实现在高中阶段的目标，尽管他们只能够在他们所教授的课程上尽力帮助我解决在课堂上遇到的问题，而我觉得这样已经足够了，因为我可以把课堂上的知识和那种研究的氛围带到我课外喜欢做的事情上，并且有信心把它做得更好！ 一个人跳舞，这是一种什么样的感觉？我想只有这样的经历才会以一种无以言表的语言给它勾画出最美的舞姿。 以前我仅在电视中看到许多人为了自己的事业和学业常常孤身一人站在舞台上跳舞。我原以为那些虚构的剧情和人物不会发生在我的身上，想不到它却悄悄地走进了我生活的每一个角落，当周围的万家灯火都熄灭入睡的时候，自己一个人还伏在实验台和课桌上忙着学习，然而我并不因为这种一个人跳舞的感觉而感到任何的不适，我觉得做任何一件事都要经历这样一个过程，尤其是正在经历这样一个过程的人，千万不要降低对自己的要求，如果放弃了，在未来的时间你会偶然发现，其实成功离我们很近，如果在那个时候我只要不放弃，我就可以实现自己的价值了。同样，这种一个人跳舞的感觉在很大程度上考验着我们的毅力和恒心，也加深了我对我未来的职业理解，为什么呢？ 我们每个人都知道，当你把你的兴趣和爱好融入到学习、工作任何一个角落的时候，并且当你第一次想用自己的兴趣额和爱好的实现您的价值时候，你却感到无比的茫然。举一个典型的例子：做初期的技术开发是非常困难的，无论从商业、学习以及个人的主观角度的理解来说，这是一个既新鲜又好玩的东西，但也充满风险和挑战，而如今我却面临这样一个十字路口，但我选择的是追随自己的兴趣和爱好，继续做自己喜欢的技术开发，对于我来说，在这个过程中的点点滴滴都值得我在将来的时间中回味… 当你一个人跳舞快跳完的时候，你就会倍加珍惜你所在的团队，而你的团队将会和你一起跳舞，而这种舞就是人们最愿意跳的舞——快乐！

回复
实
2010年01月29日 21:21 编辑
哦，我后来用那个获取鼠标状态的vi实现了滚动鼠标滚轮，改变滚动条的值：）

回复
Qizhen
2010年01月29日 10:26 编辑
似乎不能。

回复
实
2010年01月26日 14:03 编辑
再来请教一个问题，LV中的事件结构能否响应鼠标滚轮事件？:)

回复
实
2010年01月21日 18:48 编辑
恩，谢谢，我今天中午自己把这方法给凑了出来，我之前是把前面板显示放在运行之前，所以会闪，呵呵

回复
Qizhen
2010年01月21日 14:36 编辑
打开VI引用时，不要显示子VI（子VI不要设置为运行时显示）。动态调用时子VI运行后，再使用VI 的 Panel Open 方法显示其前面板。

回复
实
2010年01月20日 18:08 编辑
阮大侠，你好，我现在也是用这种方法调用子vi的，我做的界面都是最大化设置，并且只有标题栏，但是每次子vi界面弹出来的时候，都会闪一下，我想请教下，有没有什么方法让界面在弹出来时不要闪，或者不要闪的太明显？谢谢！！！

回复


1

**Reply by ruanqizhen** [reply]
之前博客下的留言：

ruanqizhen
2011年03月29日 10:55 编辑
我之前有个这相发，觉得是可行的，也没有真正给它写代码。既然你提出来需要一个范例，我今天就抽空做了一个简单的例子。我会把这个例子的介绍就添加在原文的下面。

回复
zyh7148
2011年03月29日 13:29 编辑
阮工：
你好！你给的例程是通过在界面VI中调用代码VI实现的，如果界面上的控件比较多的话，那么传给代码VI的输入参数就比较多，有没有方法不通过直接调用代码VI，而是通过控件名来传值，比如通过VI的Set Value或Get Value方法等，或有更好的方法。谢谢！

回复
ruanqizhen
2011年03月30日 08:47 编辑
你可以把界面VI的引用传给Main.vi，然后再Main.vi中通过控件名就可以得到对应控件的引用，之后就可以通过来做赋值、事件响应等的操作了。
你可以参考这篇文章里关于得到控件引用的部分：https://ruanqizhen.wordpress.com/2008/07/10/%E7%95%8C%E9%9D%A2%E8%AE%BE%E8%AE%A1%E6%8A%80%E5%B7%A7-1-%E5%88%A9%E7%94%A8-labview-%E8%87%AA%E5%B8%A6%E6%8E%A7%E4%BB%B6/
或者你有我的书的话，对应的内容在6.2.3和10.3.2节中。

回复
zyh7148
2011年03月30日 13:44 编辑
阮工：
你好！那如果通过这种方法，那在主VI中注册动态事件时，只能注册面板中所有Control Value Change事件，就不能像你例程中那样对每个控件注册一个动态事件了，只能在事件结构中再判断是哪个控件发生了变化，是这样吗？顺便问一下，在LabVIEW的安装目录下有哪几个目录下的VI是非常有用，但在程序面板中没有显示出来的？就像你的一个例程中Get Control这样的VI。你对LabVIEW的文件结构可能更了解。非常感谢！

回复
陈小刚
2011年03月30日 16:37 编辑
我喜欢把界面上的控件组合成一个一个的簇来存放，这些簇全都用严格类型定义。把这些簇的引用传入子vi再去做处理，跟用Get Control差不多，不过会稍微好用一点，毕竟你自己做过分门别类，不知道这个用法怎么样。

回复
ruanqizhen
2011年03月31日 09:25 编辑
我的意思是把界面VI的引用作为参数。再通过VI的引用，得到VI上面每个控件的引用。在Main.vi中得到Interface.vi上控件的引用之后，就既可以使用这个引用注册事件，也可以使用这个引用修改属性值了。

回复
zyh7148
2011年03月31日 11:18 编辑
阮工：
你好！可能我还是没有说明我的意思，我的意思是你在interface.vi中是把Main.vi作为子vi调用的，你的例程中interface.vi上是三个控件，你的Main.vi的输入参数就是三个，因为Main.vi中你注册动态事件时是对每个控件的引用分别注册动态事件的，因为对每个控件事件的处理是不一样的，如果interface.vi上的控件比较多时，Main.vi就要有很多输入参数，我的意思是能否有一种方法，当interface.vi上的控件比较多时，Main.vi的输入参数比较少，但又能实现你例程的功能。谢谢！

回复
ruanqizhen
2011年03月31日 13:32 编辑
我的意思是：在interface.vi与Main.vi之间，只有一个参数，是interface.vi的引用。
在Main.vi中通过这一个引用就可以得到interface.vi上所有的控件。
建议你看一下我发给你的链接。

回复
陈小刚
2011年03月31日 17:28 编辑
对，是这样，阮老师是从宏观的角度，我的角度比较微观一点:)。我的意思是，在Interface1和2里，如果控件数量很多，如果不用簇来规整的话，直接去Get Control，注册事件的时候就会非常多，比如例子里是2个按钮，就是注册两个value change，如果是10个的话就得下拉10个。如果把这10个按钮组成一个簇，注册的时候只需一次，事件处理的时候，可以写在一个事件里（cluster-All Element)。如果用簇把控件组合成几块，在Main.vi里注册和处理事件的时候就会相对少一点。另外，再把这些簇的引用做成type definition,如果interface上的控件有更改的话，修改起来也比较方便。

回复
zyh7148
2011年03月28日 17:11 编辑
阮工：
你好！我想要利用这个功能，能否给一个简单的例子，尤其是真正执行代码的那个VI，多谢了！从界面VI我是利用了VI的Set Value和Get Value方法，虽然我了解了一些动态注册事件，但代码VI我还是不知怎样去实现。

回复
ruanqizhen
2011年02月21日 12:29 编辑
使用属性节点的确比直接读写值效率低下。但是，界面不同于数学运算，在界面操作中，效率的瓶颈是人的反应速度，而不是计算机的云算速度。一个界面，假设有上千个控件，全部使用属性节点，可能效率最多低几个毫秒，这几个毫秒，人眼根本无法察觉。

回复
huangchen
2011年02月19日 23:33 编辑
阮大侠，这样做的话，因为只有控件的引用，读取和赋值都得通过属性节点，岂不是程序效率会大大降低？

回复
